### ISSUE TITLE
SSH via Putty Agent

### ISSUE THREAD

**Issue by gnunes76** [ForkIssues/TrackerWin#179]
Hi. First of all, I'd like to congratulate your great application.

Is it possible for you to create some bridge to access ssh keys in Windows putty agent?

Thank you.

**Comment by DanPristupov** [comment]
I don't really know much about Putty. Does it have a different ssh key format?

It looks like you can convert its keys to Unix format: https://stackoverflow.com/questions/2224066/how-to-convert-ssh-keypairs-generated-using-puttygen-windows-into-key-pairs-us

Update: wrong. You want to use the private key, not public one.

**Comment by mikebeaton** [comment]
Hi - the standard approach, which Git for Windows (GfW) and FileZilla (for instance) both use [EDIT: and also VS and VS Code - which I think I'm right in saying is because they use GfW], is to integrate with Pageant (which is the PuTTY equivalent of ssh-agent). While Pageant is running, then programs which know about it can get the relevant keys from it and use them to authenticate.

This means I just need to start Pageant once, and then many different programs will authenticate using it - again very much like ssh-agent.

PS While you're there, yet another new(ish) system in Windows which it might be worth being aware of is WSL, which *does* have an instance of ssh-agent which can (with effort!) be configured to start once the first time you start the WSL Ubuntu Bash shell and then keep running. Git in WSL Ubuntu (which is NOT the same version of git as Git for Windows, just to be clear) knows about that version of ssh-agent and uses it if it is running, but does not know about Pageant! FWIW and IMHO it would be more helpful to more people to provide Pageant integration, though. 😉 

I hope that all helps!

**Comment by mikebeaton** [comment]
Sorry, I hadn't even tried pushing with Fork prior to the above comment.

@gnunes76 this already works!

Just open your PuTTY key with Pageant (download Pagent from the PuTTY distro if you don't already have it) and leave that program running. Fork is (presumably?!) already using Git for Windows, and this just works!

(If you're already doing this and it's not working for you, I'd suggest double-checking that it's even working using the GfW version of `git` from the command line? It should, if all is set up right.)

**Comment by DanPristupov** [comment]
To be honest I'm really surprised you made it work so easily. Please share your experience later.

Since WSL was mentioned, we have a similar problem with it: https://github.com/ForkIssues/TrackerWin/issues/160#issuecomment-464517051

**Comment by mikebeaton** [comment]
**My setup (which I am 99% sure is completely normal!):**

If you use Fork try to pull from an ssh enabled remote ([git@bitbucket.org:[user]/[repo].git]() or [git@github.com:[user]/[repo].git]()) without PuTTY Pageant started (and without any SSH keys set up in Fork) then you get the error: `FATAL ERROR: Disconnected: No supported authentication methods available (server sent: publickey)`.

**However, if you pull from (or otherwise access) an ssh enabled git remote using Fork while PuTTY Pageant is running, but still with no SSH keys set up in Fork, then... it just works!** (As it should do: Fork uses GfW and GfW already uses Pageant if available.) [EDIT: False! But it does all 'just work' as long as you set up your `GIT_SSH` variable, see bottom of this post.]

**How to set up Pageant?**

Once only:
 - Download the full install of PuTTY, including puttygen, pageant, etc. (if you don't have it)
 - Start `puttygen`
 - Click 'Load'
 - Navigate to your existing private key file; you will need to select file type 'All Files' to see it
 - Open it (you will need its password)
 - Click 'Save private key' to save a PuTTY .ppk file containing your private key

Then, once on every login to Windows:
 - Double-click on your new .ppk file; this should start Pageant
 - Enter the private key's password
 - You should see the icon for Pageant in the taskbar icons area

While Pageant is running this private key file will be automatically used in:
- FileZilla
- Git for Windows [EDIT: as long as the `GIT_SSH` environment variable is set up! See edit at bottom.]
- Anything which uses GfW, including:
    - Visual Studio
    - Visual Studio Code
    - Fork (!)

(PS Inside the `Fork` distro is a folder called `gitInstance` which currently contains GfW 2.20.1, so I'm guessing that Fork always uses that rather then the user's own install of GfW, even if they have one?)

**EDIT:**

Oh dear, crucial item which I forget to include (because I already had it from ages ago and couldn't see that it was missing until I came to re-do this on brand new machine):

You must set the environment variable:

```
GIT_SSH=C:\Program Files\PuTTY\plink.exe
```

(https://stackoverflow.com/a/43313491/795690)

**Comment by MartinLechner-TD** [comment]
I had the same problem with the missing GIT_SSH environment variable. Interesting was that other tools (GitExtension, Sourcetree) which I used worked without these environment variable.

**Comment by HighPriest** [comment]
I would also like to bring your attention that
 - Enabling any key in Forks built in key manager disables ssh-agents completely

**Comment by milosa** [comment]
> I would also like to bring your attention that
> 
> * Enabling any key in Forks built in key manager disables ssh-agents completely

I had to go to `C:\Users\<username>\AppData\Local\Fork\settings.json` and delete the `SshKeys` field of the JSON object completely to reset that. Now it takes the key from Pageant.

**Comment by DanPristupov** [comment]
@milosa you could uncheck it in the `File -> Configure SSH keys` dialog

**Comment by mikebeaton** [comment]
See https://github.com/fork-dev/TrackerWin/issues/160#issuecomment-878500203 for a v. useful project which allows WSL command line `ssh` (and command line `git` for that matter) to also use keys from Pageant -  so one place to store keys for everything! :)

**Comment by MichalBrix** [comment]
> [@milosa](https://github.com/milosa) you could uncheck it in the `File -> Configure SSH keys` dialog

No, you can't, or at least I didn't know how to do it if you don't have any ssh keys defined. I had this field as an empty array with nothing in it. Push was failing. I had to delete that field from settings.json - now putty key is working.
